### DISCUSSION TITLE
Metals 0.11.1 Queries & Observations

### DISCUSSION THREAD

**Discussion by vishalovercome** [scalameta/metals#3535] [General]
1. Is this something to worry about - `ERROR ERROR StatusLogger Reconfiguration failed: No configuration found for '250b5e5b' at 'null' in 'null'`
2. Running tests with test explorer doesn't seem to be working. Worse, it seems to be stuck at "Running tests, 0/1 passed (0.00%)". I also see this stray process (jps output): `15431  -- main class information unavailable`
3. Output of jstack (when idle) attached - [jstack.txt](https://github.com/scalameta/metals/files/7881991/jstack.txt)
4. CPU has dropped but performance is still worse than 0.10.x
5. Can't run any more tests once I hit (2)
6. Started getting https://github.com/scalameta/metals/issues/3532 with the new release

**Comment by kpodsiad** [comment]
Thanks for another portion of feedback @vishalovercome.
2. Could you switch to code lenses and check if the problem is still present?

**Reply by vishalovercome** [reply]
I deleted all the generated files and did a fresh build and it works now. There are tickets for some open issues and I'd like to add one more to it - Java tests aren't discovered currently.

**Reply by kpodsiad** [reply]
Which framework is used in your Java tests?

**Reply by vishalovercome** [reply]
JUnit 5

**Reply by Arthurm1** [reply]
JUnit 5 isn't supported by bloop yet: https://github.com/scalacenter/bloop/issues/996
The issue is with test discovery

**Comment by tgodzik** [comment]
I converted this into discussion, since this isn't exactly one issue. Let's try to open issue related to one particular problem that we can fix.

> CPU has dropped but performance is still worse than 0.10.

Could you explain a bit more about it? Is it the compialtion taking more time? Maybe it's more memory being used with Java files added and we could increase Xmx?

**Comment by vishalovercome** [comment]
@Arthurm1 can you shed light on the next steps on Java support? In an another thread you had said that disabling Java support would not be easy. These are the current problems that need to be solved:

1. Metals detects spurious errors in Java code which wasn't the case with RedHat Java Plugin
2. Multiple compilations are needed before these "errors" go away.
3. Java tests aren't detected by the test explorer API
4. Performance is still an issue even after the fixes put in - auto-completion is barely usable

**Comment by Arthurm1** [comment]
@vishalovercome Thanks for all the feedback.

In terms of new functionality - that's undecided.  But I'm certainly looking at the various issues that have been raised.

> 1. Metals detects spurious errors in Java code which wasn't the case with RedHat Java Plugin
> 2. Multiple compilations are needed before these "errors" go away.

This shouldn't happen - Bloop should happily compile any java code that SBT has compiled.  The only thing I can think of is that incremental compilation has an issue with a certain set of changes.  Does this happen after issuing a `clean compile workspace` command from Metals?  Or only after working on a file for a while?  Do you have a way to reproduce it and could you supply that?

I'd like to know more about your project setup i.e. how many projects Metals lists and do you have scala code that depends on java source or vice versa or mixed.  What version of Java are you targeting (seems like JDK 11 from your stacktraces?)  Also - how were you using the RedHat Java plugin in combination with Metals before we added java functionality - was it with pure java projects - if not, how were you getting it to compile Scala?

> 3. Java tests aren't detected by the test explorer API

Is this the same as you've written above `Can't run any more tests once I hit (2)`?  Or is this that certain tests aren't being recognised at all?  If it's the 2nd - what test framework are you using?  There is an issue with JUnit 5 which is currently not supported by Bloop.

> 4. Performance is still an issue even after the fixes put in - auto-completion is barely usable

This is for Java files or Scala files?  I can try and reproduce this but do need to know how you've setup your project files to work with the java plugin. Currently we respond to VScode's `textDocument/completion` request virtually immediately with a blank list for java files so VSCode should just respond with the java plugin's completion list but obviously there's some issue with this.

**Reply by tgodzik** [reply]
> Metals detects spurious errors in Java code which wasn't the case with RedHat Java Plugin

This hasn't changed with 0.11.1, the same errors would be reported previously. That said, there aren't usually a lot of problems with Java classfiles since it's simpler to detect the changes.

> Java tests aren't detected by the test explorer API

I think that's because Bloop doesn't report them, but I am not 100% sure about it still.

> Performance is still an issue even after the fixes put in - auto-completion is barely usable

As @Arthurm1 said, it would be good to have any kind of reproduction and some more information. I haven't found any new issues while working with 0.11.1 in various projects.

**Reply by vishalovercome** [reply]
1. Its a multi-project repo with 3 sub-projects. 
2. Metals ends up detecting 10 projects (main, test, it)
3. Our code relies on various frameworks such as flink, kafka, dynamo, mongodb etc. and their codebase is a mix of Scala and Java
4. My own codebase is written in a way that Scala code can invoke Java code but not the other way round
5. Given (3) I would imagine that the project as a whole doesn't follow any rule on what calls what
6. Code is in Scala 2.12.11 + Java 11 (adoptopenjdk)
7. Previously I did what was recommended - run RedHat plugin for Java support and link it with Scala using sbt-eclipse. It was _fine_ for what it did. What was missing was - debugging support and refactoring of Java code resulting in update to Scala code (if needed). Resource usage of that plugin was very low as well, which wasn't the case for bloop/sbt earlier as well
8. Metals + sbt-eclipse worked and resource utilization didn't affect auto-completion for instance. Now auto-completion is by and large unusable. 
9. Besides high CPU, memory usage has gone up as well. I have to restart bloop at least 2 times a day
10. Spurious warnings and errors are of two categories:
	1. Warnings that can never go away - Most of them are in my Java unit tests (JUnit5 based)
	2. Errors in main source code that take a lot of time before they go away.
11. My suspicion is that Metals doesn't do a good job with Java annotations. Even with RedHat plugin, I had to install a [Lombok plugin](https://marketplace.visualstudio.com/items?itemName=GabrielBB.vscode-lombok). Most of the spurious errors are about getter/setter methods that Metals believes doesn't exist when in fact they do. As is the nature of Java projects, there are various other annotations from other libraries as well
12. I've noticed a bizarre error when renaming Java classes these days - sometimes it ends up modifying the top-most Lombok annotations in some other file where the class was being used/defined. 
13. Java tests are in JUnit5. Metals cannot detect them and RedHat plugin cannot run them as it crashes while trying to run them (as it can't find the classes for some reason)

**Reply by vishalovercome** [reply]
In the Java support placeholder ticket you've acknowledged that you can never support Java as well as RedHat + Microsoft will. This was obvious to me from the get go, which is why what I wanted as part of Java support are:

1. Proper debugger support - watch, conditional breakpoints, logpoints should _just work_ when they're practically unusable. Please take a look at how well IntelliJ works in this regard.
2. Refactoring of Java code should result in change to Scala code as well. Luckily my code is written in a way that minimizes Scala - Java interaction. Otherwise it would have been a nightmare.

Things which I feel are unnecessary:

1. Code formatter
2. Renaming / refactoring Java code via metals - maybe its needed for refactoring Scala + Java code as whole
3. Anything that can be offloaded to RedHat plugin

A general rule of thumb would be to explore how we can use Microsoft+RedHat efforts rather than redoing them. We have plenty of issue with bloop/vscode itself and should focus on making that work

**Reply by tgodzik** [reply]
> 8. Metals + sbt-eclipse worked and resource utilization didn't affect auto-completion for instance. Now auto-completion is by and large unusable.

Are you sure you are on the latest Metals version when testing? `0.11.1` ? We haven't been getting any other reports about degraded performance after the recent fix.

> 9. Besides high CPU, memory usage has gone up as well. I have to restart bloop at least 2 times a day

For larger projects it's best to setup `-Xmx` for Bloop otherwise it will take up any memory available. There were no changes to Bloop in that regard recently.

> 10 - 11

I think Bloop doesn't work well with annotations and code generation, that's true. But again, that would not have worked before `0.11.1` either.

> 12. I've noticed a bizarre error when renaming Java classes these days - sometimes it ends up modifying the top-most Lombok annotations in some other file where the class was being used/defined.

That's an error, if you are able to reproduce it in a smaller setting, please do report it.

Overall, it seems that a Java heavy workspace will work better with Intellij, which has a much more mature Java support and I don't think Metals will ever provide.

**Reply by tgodzik** [reply]
> A general rule of thumb would be to explore how we can use Microsoft+RedHat efforts rather than redoing them. We have plenty of issue with bloop/vscode itself and should focus on making that work

Microsoft+RedHat does not work at all with Scala, it's unable to compile it even in a mixed workspace, which is why we wanted to include Java support as part of Metals. We do use the same formatter and scanner. We also tried to include JDT as part of Metals as does the Redhat plugin, but that requires OSGI and is quite complex to set up. The biggest problem here is that there is no actual standard way to interact between servers.

> Proper debugger support - watch, conditional breakpoints, logpoints should just work when they're practically unusable. Please take a look at how well IntelliJ works in this regard.

Expression evaluation in Java is not supported, which is why that does not work currently. Intellij is a great tool, which already works well with Java, so as I said previously, it is recommended for Java heavy workspaces. Metals is about Scala + some Java files and we don't aim to be more than that, since this would be incredibly hard to do.

**Reply by vishalovercome** [reply]
Thanks a lot for such a quick response!

1. I'm using the latest plugin and definitely performance degradation has happened. Its not as bad as 0.11 but 0.11.1 isn't as good as 0.10.x. Maybe my project is one of the few with more Java than Scala code resulting in me experiencing more of a problem but I can assure you that its there
2. Editing Java source code has become extremely painful - I use Error lens plugin that overlays error/warning messages within the source code and all I see is a sea of red because Metals takes too long to process changes. Could be just a matter of annotations. 

I understand that you can't provide as good a Java support as Scala support. In that case you should offer feature flags for various Java related features. If debugger support cannot be enhanced then I would prefer 0.10.x + Test Explorer + New Scala Features. I can't do that right now as I can't disable new Java support. Definitely flags for disabling code formatting and renaming and other "on the fly" tasks should be considered as they're most likely the reason for drop in performance. If it's not possible then let me know. 

Many thanks for your support!

**Reply by tgodzik** [reply]
I don't think disabling is an option, we try to keep the settings to the minimum. It would be good however to figure out what is causing the preformance degradation for you. 

Are you able to see what is taking up the bulk of the CPU usage? If it's Bloop then it's a compilation issue, but if it's Metals it might be related to formatting or tokenizing. Would you be able to take a stack trace of the process causing your performance issue?

**Reply by Arthurm1** [reply]
Just for your info - the formatting only kicks in when you use the VSCode `Format Document` and `Format Selection`.  Metals doesn't support `Format on type` for Java.  The same with renames - they don't use CPU until they're told to by the user.

**Comment by vishalovercome** [comment]
Would you recommend a JDK other than AdoptOpenJDK? I've read that Zulu JDK results in much better IntelliJ performance but I've never seen any such recommendations in Metals/Bloop doc

**Reply by tgodzik** [reply]
No idea, but we've seen some issues with some Zulu versions in Metals, though it might have been an isolated incident.

**Comment by Arthurm1** [comment]
@vishalovercome I've had a play with using SBT + Lombok + Java + Scala + Metals.  Metals seems to handle compilation (in java + scala) and code completion (in scala) ok.

A simple Java class...
```java
package foo;

import lombok.Data;

@Data
class Foo {
  public String bar;
}
```

gets compiled to...
```java
class Foo {
    public String bar;

    public String getBar() {
        return this.bar;
    }

    public void setBar(String string) {
        this.bar = string;
    }

    public boolean equals(Object object) {
        if (object == this) {
            return true;
        }
        if (!(object instanceof Foo)) {
            return false;
        }
        Foo foo = (Foo)object;
        if (!foo.canEqual(this)) {
            return false;
        }
        String string = this.getBar();
        String string2 = foo.getBar();
        return !(string == null ? string2 != null : !string.equals(string2));
    }

    protected boolean canEqual(Object object) {
        return object instanceof Foo;
    }

    public int hashCode() {
        int n = 1;
        String string = this.getBar();
        n = n * 59 + (string == null ? 43 : string.hashCode());
        return n;
    }

    public String toString() {
        return "Foo(bar=" + this.getBar() + ")";
    }
}
```

and those extra methods can be used in Scala in the same project.  This compiles fine...
```scala
package foo

object BarGetter {
  val aFoo = new foo.Foo()
  aFoo.getBar()
}
```

You can't navigate to `getBar()` (as it's not in the java source, only the class file) but it does come up in code completion.

My SBT project is very simple.  I'm using Java 17, hence the additional javac options...
```

libraryDependencies += "org.projectlombok" % "lombok" % "1.18.16"
fork := true
javacOptions in (Compile, compile) ++= Seq(
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED",
   "-J--add-opens=jdk.compiler/com.sun.tools.javac.jvm=ALL-UNNAMED"
)
```

Could you see if a you could setup a small project that reproduces your issues.  Maybe you have some specific build settings that are causing annotation processing issues within bloop.

**Reply by vishalovercome** [reply]
Does this help - https://github.com/scalameta/metals/discussions/3562#discussioncomment-2033818 ? Reproducing with a smaller project will take me some time - probably I'll get to it this weekend.

**Reply by vishalovercome** [reply]
I don't use any of the Java options that you have mentioned above. Only bit of Java configuration is:

`compileOrder := CompileOrder.JavaThenScala`

**Reply by vishalovercome** [reply]
FWIW - code navigation works for getters and setters. It always has.

**Reply by vishalovercome** [reply]
![Screenshot 2022-01-24 at 6 59 14 PM](https://user-images.githubusercontent.com/50405519/150791535-6d563700-3b0c-4811-8ee4-798bf9cbd778.png)

These three lines have 50+ compiler errors according to Metals. Previously I was using just bloop + Metals and this error would come and go away with 0.11.x. It was never a problem with 0.10.x (sbt-eclipse + redhat). Today after Chris's recommendation I switch to sbt and now this error is _permanent_. Thus my project is now broken as I can't run any launch.json tasks. I will switch back to bloop and continue running sbt for proper compilation and running tests

**Reply by tgodzik** [reply]
Are you sure that error comes from sbt though? What does it say for the source (`bloop`, `sbt` or `scalameta` ?)
